File: apps/minerec/src/minerec/processing/actions.py
from __future__ import annotations

import json
import os
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from minerec.errors import RecorderError
from minerec.processing.capture import load_capture_metadata, scan_capture_events

ACTION_SCHEMA_VERSION = 1


@dataclass(frozen=True)
class ActionExtractionResult:
    output: Path
    record_count: int
    first_tick: int
    last_tick: int


def _payload(record: dict[str, Any]) -> dict[str, Any]:
    value = record.get("packet")
    if isinstance(value, dict):
        return value
    envelope = {
        "schema_version",
        "record_type",
        "session_id",
        "server_tick",
        "sequence",
        "recorded_at_ns",
        "recorded_at_unix_ms",
        "player_uuid",
        "player_name",
        "entity_id",
        "connection_id",
        "connection_start_server_tick",
        "apply_sequence",
        "arrival_sequence",
    }
    return {key: value for key, value in record.items() if key not in envelope}


def _action_type(record: dict[str, Any], payload: dict[str, Any]) -> str:
    if record.get("record_type") == "control_state":
        return "control_state"
    for key in ("action_type", "action_kind", "packet_type", "packet_class", "type"):
        value = record.get(key)
        if isinstance(value, str) and value:
            return value
        value = payload.get(key)
        if isinstance(value, str) and value:
            return value
    return "unknown_serverbound_packet"


def _action_row(record: dict[str, Any]) -> dict[str, Any]:
    payload = _payload(record)
    return {
        "schema_version": ACTION_SCHEMA_VERSION,
        "server_tick": record["server_tick"],
        "sequence": record["sequence"],
        "apply_sequence": record.get("apply_sequence"),
        "player_uuid": record["player_uuid"],
        "connection_id": record["connection_id"],
        "action_type": _action_type(record, payload),
        "payload": payload,
        "source_record_type": record["record_type"],
    }


def extract_actions(
    metadata: Path,
    events: Path,
    output: Path,
    *,
    first_tick: int | None = None,
    last_tick: int | None = None,
    force: bool = False,
) -> ActionExtractionResult:
    """Reconstruct one connection action stream from recorder events into one file."""

    capture = load_capture_metadata(metadata)
    if first_tick is not None and last_tick is not None and first_tick > last_tick:
        raise RecorderError("first tick cannot be greater than last tick")
    requested = output.expanduser()
    if requested.is_symlink():
        raise RecorderError(f"action output may not be a symlink: {requested}")
    destination = requested.resolve()
    destination.parent.mkdir(parents=True, exist_ok=True)
    if destination.exists() or destination.is_symlink():
        if not force:
            raise RecorderError(f"action output exists: {destination}; pass --force to replace it")
        if destination.is_symlink() or not destination.is_file():
            raise RecorderError(f"action output is not a replaceable regular file: {destination}")

    descriptor, staging_name = tempfile.mkstemp(prefix=f".{destination.name}.", suffix=".inprogress", dir=destination.parent)
    os.close(descriptor)
    staging = Path(staging_name)
    count = 0
    observed_ticks: list[int] = []
    try:
        with staging.open("wb") as handle:

            def visit(record: dict[str, Any]) -> None:
                nonlocal count
                if record.get("record_type") not in {"control_state", "packet_apply"}:
                    return
                tick = record["server_tick"]
                if (first_tick is not None and tick < first_tick) or (last_tick is not None and tick > last_tick):
                    return
                handle.write(
                    json.dumps(
                        _action_row(record),
                        sort_keys=True,
                        separators=(",", ":"),
                        ensure_ascii=False,
                        allow_nan=False,
                    ).encode("utf-8")
                    + b"\n"
                )
                count += 1
                observed_ticks.append(tick)

            scan_capture_events(events, capture, visit)
            handle.flush()
            os.fsync(handle.fileno())
        if not observed_ticks:
            raise RecorderError("no reconstructed actions matched the requested connection and tick range")
        os.replace(staging, destination)
    except Exception:
        staging.unlink(missing_ok=True)
        raise
    return ActionExtractionResult(destination, count, min(observed_ticks), max(observed_ticks))


__all__ = ["ACTION_SCHEMA_VERSION", "ActionExtractionResult", "extract_actions"]

File: internal/models/actions/actions.go
package actions

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"

	artifactsv1 "github.com/proj-airi/mc-play-recorder/apis/sdk/go/mc-play-recorder/artifacts/v1"
	"github.com/proj-airi/mc-play-recorder/internal/models/captures"
	"github.com/samber/do/v2"
	"google.golang.org/protobuf/encoding/protojson"
)

type Options struct {
	Metadata  string
	Events    string
	Output    string
	FromTick  *int64
	ToTick    *int64
	Overwrite bool
}

type Result struct {
	Output      string
	RecordCount uint64
	FirstTick   int64
	LastTick    int64
}

type Service struct {
	captures *captures.Service
}

func NewService(injector do.Injector) (*Service, error) {
	captureService, err := do.Invoke[*captures.Service](injector)
	if err != nil {
		return nil, err
	}
	return &Service{captures: captureService}, nil
}

func (service *Service) Extract(options Options) (Result, error) {
	metadata, err := service.captures.LoadMetadata(options.Metadata)
	if err != nil {
		return Result{}, err
	}
	if options.FromTick != nil && options.ToTick != nil && *options.FromTick > *options.ToTick {
		return Result{}, errors.New("first tick cannot be greater than last tick")
	}
	destination, err := prepareOutput(options.Output, options.Overwrite)
	if err != nil {
		return Result{}, err
	}
	temporary, err := os.CreateTemp(filepath.Dir(destination), "."+filepath.Base(destination)+".*.inprogress")
	if err != nil {
		return Result{}, fmt.Errorf("create action staging file: %w", err)
	}
	temporaryName := temporary.Name()
	defer func() { _ = os.Remove(temporaryName) }()
	if err := temporary.Chmod(0o600); err != nil {
		_ = temporary.Close()
		return Result{}, err
	}
	result := Result{Output: destination}
	_, err = service.captures.ScanEvents(options.Events, metadata, func(event captures.Event) error {
		if event.RecordType != "control_state" && event.RecordType != "packet_apply" {
			return nil
		}
		if options.FromTick != nil && event.ServerTick < *options.FromTick || options.ToTick != nil && event.ServerTick > *options.ToTick {
			return nil
		}
		action := &artifactsv1.PlayerAction{
			SchemaVersion: 1, SourceRecordType: event.RecordType, ConnectionId: event.ConnectionID,
			PlayerUuid: event.PlayerUUID, ServerTick: event.ServerTick, Sequence: event.Sequence,
		}
		if state := event.Message.GetControlState(); state != nil {
			action.Payload = &artifactsv1.PlayerAction_ControlState{ControlState: state.GetState()}
		} else if applied := event.Message.GetPacketApply(); applied != nil {
			applySequence := applied.GetApplySequence()
			action.ApplySequence = &applySequence
			packet := applied.GetPacket()
			action.Payload = &artifactsv1.PlayerAction_Packet{Packet: &artifactsv1.PacketAction{
				ActionKind: packet.GetActionKind(), Packet: packet.GetIdentity(),
			}}
		}
		encoded, err := protojson.Marshal(action)
		if err != nil {
			return fmt.Errorf("encode action ProtoJSON: %w", err)
		}
		if _, err := temporary.Write(append(encoded, '\n')); err != nil {
			return fmt.Errorf("write action JSONL: %w", err)
		}
		if result.RecordCount == 0 {
			result.FirstTick = event.ServerTick
		}
		result.LastTick = event.ServerTick
		result.RecordCount++
		return nil
	})
	if err != nil {
		_ = temporary.Close()
		return Result{}, err
	}
	if result.RecordCount == 0 {
		_ = temporary.Close()
		return Result{}, errors.New("no reconstructed actions matched the requested connection and tick range")
	}
	if err := temporary.Sync(); err != nil {
		_ = temporary.Close()
		return Result{}, fmt.Errorf("sync action output: %w", err)
	}
	if err := temporary.Close(); err != nil {
		return Result{}, fmt.Errorf("close action output: %w", err)
	}
	if err := os.Rename(temporaryName, destination); err != nil {
		return Result{}, fmt.Errorf("publish action output: %w", err)
	}
	return result, nil
}

func prepareOutput(path string, overwrite bool) (string, error) {
	destination, err := filepath.Abs(path)
	if err != nil {
		return "", fmt.Errorf("resolve action output: %w", err)
	}
	if info, err := os.Lstat(destination); err == nil {
		if !overwrite {
			return "", fmt.Errorf("action output exists: %s; pass --overwrite to replace it", destination)
		}
		if info.Mode()&os.ModeSymlink != 0 || !info.Mode().IsRegular() {
			return "", fmt.Errorf("action output is not a replaceable regular file: %s", destination)
		}
		if !ownedOutput(destination) {
			return "", fmt.Errorf("refusing to replace non-PlayerAction output: %s", destination)
		}
	} else if !errors.Is(err, os.ErrNotExist) {
		return "", fmt.Errorf("inspect action output: %w", err)
	}
	if err := os.MkdirAll(filepath.Dir(destination), 0o750); err != nil {
		return "", fmt.Errorf("create action output directory: %w", err)
	}
	return destination, nil
}

func ownedOutput(path string) bool {
	file, err := os.Open(path)
	if err != nil {
		return false
	}
	defer func() { _ = file.Close() }()
	reader := bufio.NewReaderSize(file, 1024*1024)
	count := 0
	for {
		line, readErr := reader.ReadSlice('\n')
		if len(line) == 0 && errors.Is(readErr, io.EOF) {
			return count > 0
		}
		if readErr != nil || len(line) == 0 || line[len(line)-1] != '\n' {
			return false
		}
		action := &artifactsv1.PlayerAction{}
		if protojson.Unmarshal(line[:len(line)-1], action) != nil || action.GetSchemaVersion() != 1 || action.GetConnectionId() == "" ||
			action.GetPlayerUuid() == "" || action.GetSourceRecordType() == "" || action.GetPayload() == nil {
			return false
		}
		count++
	}
}
